File: Semester2/2/task4/task4/Calc.cs
using System;

namespace Task4
{
    /// <summary>
    /// Stack Calculator
    /// </summary>
    class Calc
    {
        public Calc(InterfaceStack<double> stack)
        {
            this.stackCalc = stack;
        }

        private InterfaceStack<double> stackCalc;

        /// <summary>
        /// Perform the operation
        /// </summary>
        private double CalculateResult(char operation)
        {
            double number2 = stackCalc.Pop();
            double number1 = stackCalc.Pop();
            if (operation == '+')
                return number1 + number2;
            if (operation == '-')
                return number1 - number2;
            if (operation == '*')
                return number1 * number2;
            if (operation == '/' && number2 != 0)
                return number1 / number2;
            return default(int);
        }

        private bool IsOperation(string inputString)
        {
            return (inputString == "+" || inputString == "-" || inputString == "*" || inputString == "/");
        }

        /// <summary>
        /// Get numbers and operation, calculate and add result in stack
        /// </summary>
        /// <param name="inputString">Number or operation</param>
        public void Calculate(string inputString)
        {
            if (IsOperation(inputString))
            {
                double result = CalculateResult(Convert.ToChar(inputString));
                stackCalc.Push(result);
                return;
            }
            else
            {
                stackCalc.Push(Convert.ToDouble(inputString));
            }
        }

        /// <summary>
        /// show the result of last operation
        /// </summary>
        /// <returns>result</returns>
        public double ShowResult()
        {
            return stackCalc.Top();
        }
    }
}

File: qt/Calc/calc.cpp
#include "calc.h"
#include "myException.h"

double StackCalc::calculate(QString expression, Stack *stack)
{
    int currentPosition = 0;
    int length = expression.length();
    while (currentPosition < length)
    {
        if(isOperation(expression[currentPosition].toLatin1()))
        {
            char operation = expression[currentPosition].toLatin1();
            double number2 = stack->pop();
            double number1 = stack->pop();
            double result = calculateResult(operation, number1, number2);
            stack->push(result);
            currentPosition++;
        }

        else
        {
            QString number = "";
            while(currentPosition < length && expression[currentPosition] != ' ')
            {
                number += expression[currentPosition].toLatin1();
                currentPosition++;
            }
            stack->push(number.toDouble());
        }
        currentPosition++;
    }
    return stack->top();
}

bool StackCalc::isOperation(char value) const
{
    return (value == '+' || value == '-' || value == '*' || value == '/');
}

double StackCalc::calculateResult(char operation, double number1, double number2)
{
    if(operation == '+')
    {
        return number1 + number2;
    }

    else if(operation == '-')
    {
        return number1 - number2;
    }

    else if(operation == '*')
    {
        return number1 * number2;
    }

    else
    {
        if (number2 == 0)
        {
            throw MyException("Division by zero");
        }
        return number1 / number2;
    }
}
